### PR TITLE
feat: store last sent request in localstorage

### PR DESCRIPTION
**Pull request by ValGeorgiev**
Store the last sent request in localStorage for both Extension and Electron apps.

Clean the localStorage when user use trash button

Issue: https://github.com/FormidableLabs/urql-devtools/issues/328

https://user-images.githubusercontent.com/12056467/115358773-a9d77b80-a1c6-11eb-92e9-d76db7c1aa3b.mov

### CODE REVIEW

**Review by andyrichardson — COMMENTED** [review]
Thanks for doing this Val! 🙌 

A few quick thoughts on the implementation - let me know if you want to pair on any of this.

**Handling state**

Would it be possible to use a side effect for persisting to storage? Similarly, could we use that local storage value as the default value on mount?

This should allow us to avoid having storage-based logic throughout the app.

```tsx
// Somewhere in useRequest
const state = useState({ query: window.localStorage.getItem('query') || '', /* ... */  });

useEffect(() => {
  window.localStorage.setItem('query', state.query);
}, [state.query]);
```

**Storage mechanisms**
This one is a little less straight forward and might need some investigation.

I'm guessing the expectation for persisting the query in localStorage is for these reasons:
 - User opens devtools and enters query
 - Query is persisted to storage
 - Next time user visits devtools, previous query is there

There's a potential problem with this which is that the query is being persisted to the extensions storage, but instead to an instance of the panel on a given tab. I'm not sure what this would result in but I would guess it would lead to unexpected results.

There is an API for [browser extension storage](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage) that might be more work (likely including messaging to the background thread) but should reap expected results. 

> Note: If it does require messaging to the background thread - lets hold off until we work out whether we want to [migrate to a service worker](https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/)

For electron, it looks like local storage is fine :+1:

If we end up using differing logic for storage, a util might make sense

```ts
// src/panel/util/storage.ts
export const write = (...args) => {
  if (process.env.BUILD_ENV === "extension") {
    // ...
  }
  localStorage.write(...args)
}
``` 

 ----
 
Edit: Spoke with @ValGeorgiev; localstorage will do the job - we're fine with having the value only persist for the lifetime of the tab

**Comment by ValGeorgiev** [comment]
> Thanks for doing this Val! 🙌
> 
> A few quick thoughts on the implementation - let me know if you want to pair on any of this.
> 
> **Handling state**
> 
> Would it be possible to use a side effect for persisting to storage? Similarly, could we use that local storage value as the default value on mount?
> 
> This should allow us to avoid having storage-based logic throughout the app.
> 
> ```tsx
> // Somewhere in useRequest
> const state = useState({ query: window.localStorage.getItem('query') || '', /* ... */  });
> 
> useEffect(() => {
>   window.localStorage.setItem('query', state.query);
> }, [state.query]);
> ```
> 
> **Storage mechanisms**
> This one is a little less straight forward and might need some investigation.
> 
> I'm guessing the expectation for persisting the query in localStorage is for these reasons:
> 
> * User opens devtools and enters query
> * Query is persisted to storage
> * Next time user visits devtools, previous query is there
> 
> There's a potential problem with this which is that the query is being persisted to the extensions storage, but instead to an instance of the panel on a given tab. I'm not sure what this would result in but I would guess it would lead to unexpected results.
> 
> There is an API for [browser extension storage](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage) that might be more work (likely including messaging to the background thread) but should reap expected results.
> 
> > Note: If it does require messaging to the background thread - lets hold off until we work out whether we want to [migrate to a service worker](https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/)
> 
> For electron, it looks like local storage is fine 👍
> 
> If we end up using differing logic for storage, a util might make sense
> 
> ```ts
> // src/panel/util/storage.ts
> export const write = (...args) => {
>   if (process.env.BUILD_ENV === "extension") {
>     // ...
>   }
>   localStorage.write(...args)
> }
> ```

Thanks @andyrichardson for the detailed comment. I updated the first part. I will have a look in the storage mechanisms now.

File: src/panel/context/Request.tsx [hunk 1]
```diff
@@ -99,6 +101,14 @@ export const RequestProvider: FC = ({ children }) => {
     });
   }, []);
 
+  useEffect(() => {
+    if (!query) {
+      localStorage.removeItem("urql-last-request");
+    } else {
+      localStorage.setItem("urql-last-request", query);
+    }
```
Comment (andyrichardson) [inline src/panel/context/Request.tsx:109]: Non blocking, I think we should keep the state identical between reloads, even if it's empty

```suggestion
    localStorage.setItem("urql-last-request", query);
```

File: src/panel/context/Request.tsx [hunk 2]
```diff
@@ -33,7 +33,9 @@ export const RequestProvider: FC = ({ children }) => {
     response?: Record<string, unknown>;
     error?: Record<string, unknown>;
   }>({ fetching: false, response: undefined, error: undefined });
-  const [query, setQuery] = useState<string>();
+  const [query, setQuery] = useState<string>(
+    localStorage.getItem("urql-last-request") || ""
+  );
```
Comment (andyrichardson) [inline src/panel/context/Request.tsx:38]: There's two states for the query attribute:
 - `undefined` - no user input yet
 - `string` - user input

We need to default to undefined for [this functionality](https://github.com/FormidableLabs/urql-devtools/blob/master/src/panel/pages/request/components/Query.tsx#L62).

```suggestion
  const [query, setQuery] = useState<string | undefined>(
    localStorage.getItem("urql-last-request") || undefined
  );
```

Not sure why TS never complained about the incorrect type definition 🤷

**Review by andyrichardson — APPROVED** [review]

**Pull request merged**